local_cluster/tests/local_cluster.rs

```diff
@@ -285,6 +296,97 @@ fn test_listener_startup() {
     assert_eq!(cluster_nodes.len(), 4);
 }
 
+#[test]
+fn test_snapshots_restart_validity() {
+    let temp_dir = TempDir::new().unwrap();
+    let snapshot_path = temp_dir.path().join("bank_states");
+    let snapshot_package_output_path = temp_dir.path().join("tar");
+    let snapshot_interval_slots = 25;
+
+    // Create the snapshot directories
+    fs::create_dir_all(&snapshot_path).expect("Failed to create snapshots bank state directory");
+    fs::create_dir_all(&snapshot_package_output_path)
+        .expect("Failed to create snapshots tar directory");
+
+    // Set up the cluster with 1 snapshotting validator
+    let mut snapshot_validator_config = ValidatorConfig::default();
+    snapshot_validator_config.rpc_config.enable_fullnode_exit = true;
+    snapshot_validator_config.snapshot_config = Some(SnapshotConfig::new(
+        snapshot_path,
+        snapshot_package_output_path.clone(),
+        snapshot_interval_slots,
+    ));
+    let num_account_paths = 4;
+    let account_storage_dirs: Vec<TempDir> = (0..num_account_paths)
+        .map(|_| TempDir::new().unwrap())
+        .collect();
+    let account_storage_paths: Vec<_> = account_storage_dirs
+        .iter()
+        .map(|a| a.path().to_str().unwrap().to_string())
+        .collect();
+    let account_storage_paths = AccountsDB::format_paths(account_storage_paths);
+    snapshot_validator_config.account_paths = Some(account_storage_paths);
+
+    let config = ClusterConfig {
+        node_stakes: vec![10000],
+        cluster_lamports: 100000,
+        validator_configs: vec![snapshot_validator_config.clone()],
+        ..ClusterConfig::default()
+    };
+
+    // Create and reboot the node from snapshot `num_runs` times
+    let num_runs = 3;
+    let mut expected_balances = HashMap::new();
+    let mut cluster = LocalCluster::new(&config);
+    for _ in 0..num_runs {
+        // Push transactions to one of the nodes and confirm that transactions were
+        // forwarded to and processed.
+        trace!("Sending transactions");
+        let new_balances = cluster_tests::send_many_transactions(
+            &cluster.entry_point_info,
+            &cluster.funding_keypair,
+            10,
+            10,
+        );
+
+        expected_balances.extend(new_balances);
+
+        // Remove any snapshots made so far so that we know then next snapshot will include all of the
+        // balances
+        trace!("Clearing tar file");
+        let tar = snapshot_utils::get_snapshot_tar_path(&snapshot_package_output_path);
+        let _ = fs::remove_file(tar);
+
+        // Wait for snapshot to be made
+        let tar = snapshot_utils::get_snapshot_tar_path(&snapshot_package_output_path);
+        trace!("Waiting for tar to be generated");
+        loop {
+            if tar.exists() {
```

Comment:
@sagar-solana, this is the potentially flaky part. I can't be certain that the tar that is generated includes the transactions that were sent above, b/c the tar may have been generated from a bank before the transactions went through (SnapshotPackagerService runs asynchronously)